---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-"
)
```

## ngsjs

[ngsjs](https://github.com/JhuangLab/ngsjs) is a command line scripts library to facilitate the construction, distribution and executiton of reproducible next-generation sequencing (NGS) data analysis workflows including [snakemake](https://snakemake.readthedocs.io/en/stable/), [nextflow](https://www.nextflow.io/), [bpipe](https://github.com/ssadedin/bpipe) and [WDL](https://github.com/openwdl/wdl).

This is a experimental project exploring the simple way to construct, distribute and execute various next-generation sequencing (NGS) data analysis workflows.

Using the [Common workflow language (CWL)](https://www.commonwl.org/) to construct bioinformatics data analysis workflows is attractive and valuable for improving the performance and reproducible of biological data analysis works.

Now, the difficulties that need to be solved are list: 

- Lack of integration and unify of massive data analysis workflows.
- Lack of the unified distribution platform for various data analysis workflows (e.g. snakemake, nextflow, Galaxy, etc.). 
- Reuse of workflows language codes (e.g. commands, input and output information) on other programming platform are still slighly complicated.
- The readability and reusable will also decreased when massive Python and R codes mixed with the workflows language codes.

Here, we proposed that using [node](https://nodejs.org/en/) to distribute the bioinformatics data analysis workflows. The creation, update and upload of a node package are very simple. Well-tested and high-performance distribution tools of node packages, such as [npm](https://www.npmjs.com/) and [yarn](https://www.yarnpkg.com), are providing the service for more than 831,195 node packages.

Besides, we esbalish a framwork to integrate various data analysis workflows and command line scripts. The functional files used in this framework as shown below:

- rbashful: Built in the ngsjs package (or user custom script) to dynamically render and execute the env.toml commands and cli.yaml workflows.
- cli.yaml: Workflow controlor (e.g. bashful supported style).
- env.toml: Store the fields and values of input and output parameters; the function commands (Python, R or others) indexed by unique keys.
- others: Other workflow files and scripts (such as snakemake, nextflow workflows, scripts). The workflow files and scripts is better to upload as the node package, and the deployment functions for required extra files.

**Scripts:**

- rsession: Get ouput of `sessionInfo()` and `sessioninfo::session\_info()`

```{bash}
rsession -f 2
```

- rinstall: Install R packages and [BioInstaller](https://github.com/JhuangLab/BioInstaller) resources using `install.packages()` and R packages `devtools`, `BiocManager` and `BioInstaller`

```{bash}
rinstall -h
```

- rbashful: Using the GO program [bashful](https://github.com/wagoodman/bashful), yaml and toml and R scripts to stitch together commands and bash snippits and run them with a bit of style.

![](https://raw.githubusercontent.com/wagoodman/bashful/master/doc/demo.gif)

```{bash}
rbashful -h
```

## Requirements

- [node](https://nodejs.org/en/)
- [R](https://cran.r-project.org/)
- [GO](https://golang.org/)

### R packages

- optparse
- configr
- devtools
- BiocManager
- sessioninfo

## Installation

You need to install the [node](https://nodejs.org/en/), [R](https://cran.r-project.org/) and [GO](https://golang.org/) to run all functions of [ngsjs](https://github.com/JhuangLab/ngsjs).

```{bash eval=FALSE}
npm install -g ngsjs
# or
yarn global add ngsjs

# If you not to globaly install ngsjs 
# You need to set the PATH
echo "export NGSJS_ROOT=/path/node_modules/nodejs" >> ~/.bashrc
echo "export PATH=$PATH:${NGSJS_ROOT}/bin" >> ~/.bashrc

# Current dir is /path
npm install ngsjs
# or
yarn add ngsjs
```

## Usage

Before using the `ngsjs` command line tools, you should run the `resolve_r_deps` command to install the extra R packages used in `ngsjs` scripts. 

```{bash eval=FALSE}
# install the extra R packages used in `ngsjs` scripts
resolve_r_deps
```

### rsession

```{bash eval=FALSE}
# Print commandline help of rsession
rsession -h

# Print rsession R document (Just like ?sessionInfo in R client)
rsession -d

# Print R sessionInfo()
# The followed three lines are equivalent.
rsession
rsession -f 1
rsession -f sessionInfo

# The followed two lines are equivalent.
rsession -f 2 -e 'include_base=TRUE'
rsession -f sessioninfo::session_info -e 'include_base=TRUE'
```

### rinstall

```{bash eval=FALSE}
# Print commandline help of rinstall
rinstall -h

# Print rinstall R document (Just like ?sessionInfo in R client)
rinstall -d

# Install CRAN R package yaml (install.package)
rinstall yaml
rinstall -f 1 yaml

# Install R package ngstk from GitHub JhuangLab/ngstk (devtools::install_github)
rinstall -f 2 JhuangLab/ngstk

# Install R package ngstk from GitHub JhuangLab/ngstk (install.package)
# devtools::install_github with force = TRUE, ref = 'develop'
rinstall -f 2 -e "force = TRUE, ref = 'develop'" JhuangLab/ngstk

# Install Bioconductor package ggtree (BiocManager)
# BiocManager::install('ggtree')
rinstall -f 3 ggtree

# Install R packages (pacman)
# pacman::p_load(ggtree)
rinstall -f 4 ggtree

# Install and download BioInstaller resources

# Show all BioInstaller default resources name
rinstall -f BioInstaller::install.bioinfo -e "show.all.names=T"
rinstall -f 5 -e "show.all.names=T"

# Show ANNOVAR refgene and avsnp versions
rinstall -f BioInstaller::install.bioinfo -e "show.all.versions=T" db_annovar_refgene
rinstall -f 5 -e "show.all.versions=T" db_annovar_avsnp

# Show ANNOVAR hg19 refgene and avsnp
rinstall -f 5 -e "download.dir='/tmp/refgene', extra.list=list(buildver='hg19')" db_annovar_refgene
rinstall -f 5 -e "download.dir='/tmp/avsnp', extra.list=list(buildver='hg19')" db_annovar_avsnp
```

### rbashful

View a rbashful demo [here](https://github.com/JhuangLab/ngsjs/tests/rbashful/rnaseq_splicing).

```{r}
source_dir <- "~/Documents/repositories/ljf/github/ngsjs/tests/rbashful/rnaseq_splicing"

# View the cli.yaml
cat(paste0(readLines(sprintf("%s/cli.yaml", source_dir)), 
           collapse = "\n"), sep = "\n")

# View the env.toml
cat(paste0(readLines(sprintf("%s/env.toml", source_dir)), 
           collapse = "\n"), sep = "\n")

# View the submit.sh
cat(paste0(readLines(sprintf("%s/submit.sh", source_dir)), 
           collapse = "\n"), sep = "\n")
```
## Maintainer

[Jianfeng Li](https://github.com/Miachol)
